File: 240625/가능한 수열 중 최솟값 구하기/find-min-of-possible-series.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <cstdlib>
using namespace std;

#define NUM_LEN 3

int n;
vector<int> series;
int numbers[NUM_LEN] = {4, 5, 6};

bool IsEqual(int start1, int end1, int start2, int end2) {
    for(int i=0; i<=end1-start1; i++) {
        if(series[start1+i] != series[start2+i])
            return false;
    }

    return true;
}

bool IsPossibleSeries() {
    int len = 1;
    while(true) {
        int end1 = (int)series.size()-1, start1 = end1 -len+1;
        int end2 = start1-1, start2 = end2-len+1;

        if(start2 < 0) break;

        if(IsEqual(start1, end1, start2, end2)) return false;

        len++;
    }

    return true;
}

void FindMinSeries(int cnt) {
    if(cnt == n) {
        for(int i=0; i<series.size(); i++) {
            cout << series[i];
        }
        exit(0);
    }

    for(int i=0; i<NUM_LEN; i++) {
        series.push_back(numbers[i]);

        if(IsPossibleSeries()) {
            FindMinSeries(cnt+1);
        }

        series.pop_back();
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n;

    FindMinSeries(0);
    return 0;
}

File: 250404/가능한 수열 중 최솟값 구하기/find-min-of-possible-series.py
import sys

n = int(input())
numbers = [4, 5, 6]

series = []

def is_possible():
    length = 1

    while True:
        s1, e1 = len(series) - length, len(series)-1
        s2, e2 = s1-length, s1-1

        if s2<0:
            break

        if series[s1:e1+1] == series[s2:e2+1]:
            return False
        
        length += 1
    
    return True

def find_min_series(cnt):
    if cnt == n:
        for elem in series:
            print(elem, end = "")
        sys.exit(0)

    for number in numbers:
        series.append(number)

        if is_possible():
            find_min_series(cnt+1)
        series.pop()

find_min_series(0)